File: app/src/main/java/com/github/pwittchen/reactivesensors/app/SensorHelper.java
package com.github.pwittchen.reactivesensors.app;

import android.widget.TextView;
import com.github.pwittchen.reactivesensors.library.ReactiveSensorEvent;
import com.github.pwittchen.reactivesensors.library.ReactiveSensors;
import com.github.pwittchen.reactivesensors.library.SensorNotFoundException;
import com.github.pwittchen.reactivesensors.library.SensorsProxy;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;


import java.util.Locale;

class SensorHelper {
  private SensorsProxy reactiveSensors;
  private int sensorType;
  private String sensorName;
  private TextView textViewForMessage;

  SensorHelper(ReactiveSensors sensors, int type, String name, TextView textViewForMessage) {
    this.reactiveSensors = sensors;
    this.sensorType = type;
    this.sensorName = name;
    this.textViewForMessage = textViewForMessage;
  }

  Disposable createSubscription() {

    return reactiveSensors.observeSensor(sensorType)
        .subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread())
        .filter(ReactiveSensorEvent::sensorChanged)
        .subscribe(event -> {
          float x = event.sensorValues()[0];
          float y = event.sensorValues()[1];
          float z = event.sensorValues()[2];

          final String format = "%s readings:\n x = %f\n y = %f\n z = %f";
          String message = String.format(Locale.getDefault(), format, event.sensorName(), x, y, z);
          textViewForMessage.setText(message);
        }, throwable -> {
          if (throwable instanceof SensorNotFoundException) {
            textViewForMessage.setText("Sorry, your device doesn't have required sensor.");
          }
        });
  }

  void safelyDispose(Disposable disposable) {
    if (!reactiveSensors.hasSensor(sensorType)) {
      return;
    }

    if (disposable != null && !disposable.isDisposed()) {
      disposable.dispose();
    }
  }
}

File: app-kotlin/src/main/kotlin/com/github/pwittchen/reactivesensors/kotlinapp/SensorHelper.kt
package com.github.pwittchen.reactivesensors.kotlinapp

import android.widget.TextView
import com.github.pwittchen.reactivesensors.library.ReactiveSensorEvent
import com.github.pwittchen.reactivesensors.library.ReactiveSensorFilter
import com.github.pwittchen.reactivesensors.library.ReactiveSensors
import com.github.pwittchen.reactivesensors.library.SensorNotFoundException
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import org.reactivestreams.Subscriber

class SensorHelper(private val reactiveSensors: ReactiveSensors, private val sensorType: Int,
    private val sensorName: String, private val textView: TextView) {

  fun createSubscription(): Disposable {
    return reactiveSensors.observeSensor(sensorType)
        .subscribeOn(Schedulers.computation())
        .filter(ReactiveSensorFilter.filterSensorChanged())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({ reactiveSensorEvent ->
          run {
            if (reactiveSensorEvent != null) {
              val event = reactiveSensorEvent.sensorEvent
              val x = event.values[0]
              val y = event.values[1]
              val z = event.values[2]
              val message = "%s readings:\n x = %f\n y = %f\n z = %f"
              textView.text = message.format(sensorName, x, y, z)
            }
          }
        })

    /*

    (object : Subscriber<ReactiveSensorEvent>() {
          override fun onCompleted() {
          }

          override fun onError(throwable: Throwable?) {
            if (throwable is SensorNotFoundException) {
              textView.text = "Sorry, your device doesn't have required sensor."
            }
          }

          override fun onNext(reactiveSensorEvent: ReactiveSensorEvent?) {
            if (reactiveSensorEvent != null) {
              val event = reactiveSensorEvent.sensorEvent
              val x = event.values[0]
              val y = event.values[1]
              val z = event.values[2]
              val message = "%s readings:\n x = %f\n y = %f\n z = %f"
              textView.text = message.format(sensorName, x, y, z)
            }
          }
        }

     */

  }

  fun safelyDispose(subscription: Disposable?) {
    if (!reactiveSensors.hasSensor(sensorType)) {
      return
    }

    if (subscription != null && !subscription.isDisposed) {
      subscription.dispose()
    }
  }
}
